File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/mysqlsource/MysqlSourceOpExec.java
package edu.uci.ics.texera.workflow.operators.mysqlsource;

import edu.uci.ics.texera.workflow.common.operators.source.SourceOperatorExecutor;
import edu.uci.ics.texera.workflow.common.tuple.Tuple;
import edu.uci.ics.texera.workflow.common.tuple.schema.Attribute;
import edu.uci.ics.texera.workflow.common.tuple.schema.AttributeType;
import edu.uci.ics.texera.workflow.common.tuple.schema.Schema;
import scala.collection.Iterator;

import java.sql.*;
import java.util.HashSet;

public class MysqlSourceOpExec implements SourceOperatorExecutor {
    private final Schema schema;
    private final String host;
    private final String port;
    private final String database;
    private final String table;
    private final String username;
    private final String password;

    private final String column;
    private final String keywords;
    private final Boolean progressive;
    private final Long interval;

    private Connection connection;
    private PreparedStatement currentPreparedStatement;
    private ResultSet resultSet;
    private final HashSet<String> tableNames;
    private boolean queriesPrepared = false;

    private Long currentLimit;
    private Long currentOffset;
    private Number currentMin = 0;
    private Number max = 0;
    private Attribute batchByAttribute = null;

    private Tuple cachedTuple = null;

    MysqlSourceOpExec(Schema schema, String host, String port, String database, String table, String username,
                      String password, Long limit, Long offset, String column, String keywords,
                      Boolean progressive, String batchByColumn, Long interval) {
        this.schema = schema;
        this.host = host.trim();
        this.port = port.trim();
        this.database = database.trim();
        this.table = table.trim();
        this.username = username.trim();
        this.password = password;
        this.currentLimit = limit;
        this.currentOffset = offset;
        this.column = column == null ? null : column.trim();
        this.keywords = keywords == null ? null : keywords.trim();
        this.progressive = progressive;
        this.interval = interval;
        this.tableNames = new HashSet<>();
        if (batchByColumn != null) {
            this.batchByAttribute = schema.getAttribute(batchByColumn);
        }

    }

    /**
     * @return A iterator of Texera.Tuple
     */
    @Override
    public Iterator<Tuple> produceTexeraTuple() {
        return new Iterator<Tuple>() {
            /**
             * check if query is sent to mysql server and hasNext flag is true
             * @return bool
             */
            @Override
            public boolean hasNext() {

                // if existing Tuple in cache, means there exist next Tuple.
                if (cachedTuple != null) {
                    return true;
                }

                // cache the next Tuple
                cachedTuple = next();

                return cachedTuple != null;
            }

            /**
             * get the next row from the resultSet and parse it into Texera Tuple
             * if there is no more in resultSet, set the hasNext flag to false and return null
             * otherwise, base on the the schema given (which is generated in MysqlSourceOpDesc.querySchema())
             * extract data from resultSet and add to tupleBuilder to construct a Texera.tuple
             *
             * @return Texera.Tuple
             */
            @Override
            public Tuple next() {

                // if has the next Tuple in cache, return it and clear the cache.
                if (cachedTuple != null) {
                    Tuple tuple = cachedTuple;
                    cachedTuple = null;
                    return tuple;
                }

                // otherwise, send query for next Tuple.
                try {
                    if (resultSet != null && resultSet.next()) {
                        if (currentOffset != null && currentOffset > 0) {
                            currentOffset--;
                            return next();
                        }
                        Tuple.Builder tupleBuilder = Tuple.newBuilder();
                        for (Attribute attr : schema.getAttributes()) {
                            String columnName = attr.getName();
                            AttributeType columnType = attr.getType();
                            String value = resultSet.getString(columnName);
                            if (value == null) {
                                tupleBuilder.add(attr, null);
                                continue;
                            }
                            switch (columnType) {
                                case INTEGER:
                                    tupleBuilder.add(attr, Integer.valueOf(value));
                                    break;
                                case LONG:
                                    tupleBuilder.add(attr, Long.valueOf(value));
                                    break;
                                case DOUBLE:
                                    tupleBuilder.add(attr, Double.valueOf(value));
                                    break;
                                case BOOLEAN:
                                    tupleBuilder.add(attr, !value.equals("0"));
                                    break;
                                case STRING:
                                    tupleBuilder.add(attr, value);
                                    break;
                                case TIMESTAMP:
                                    tupleBuilder.add(attr, Timestamp.valueOf(value));
                                    break;
                                case ANY:
                                default:
                                    throw new RuntimeException("MySQL Source: unhandled attribute type: " + columnType);
                            }
                        }
                        if (currentLimit != null) {
                            currentLimit--;
                        }
                        return tupleBuilder.build();
                    } else {
                        if (resultSet != null) {
                            resultSet.close();
                        }
                        if (currentPreparedStatement != null) {
                            currentPreparedStatement.close();
                        }

                        currentPreparedStatement = getNextQuery();
                        if (currentPreparedStatement != null) {
                            resultSet = currentPreparedStatement.executeQuery();
                            return next();
                        } else {
                            return null;
                        }

                    }

                } catch (SQLException e) {
                    throw new RuntimeException(e);
                }
            }
        };
    }

    /**
     * during the open process, a connection is established with the remote mysql server
     * then an sql query is generated using the info provided by the user.
     * A prepared statement is used to prevent sql injection attacks
     * Since user might provide info in a combination of column, keywords, limit and offset
     * the prepared statement can have different number of parameters.
     * A variable curIndex is used to keep track of the next parameter should be filled in
     */
    @Override
    public void open() {
        try {
            if (!queriesPrepared) {
                connection = this.establishConn();

                // load user table names from the given database
                loadTableNames();

                if (!tableNames.contains(table)) {
                    throw new RuntimeException("MysqlSource can't find the given table `" + table + "`.");
                }

                if (progressive) {
                    // load for batch statistics used to split mini queries
                    loadBatchColumnStats();
                }
                queriesPrepared = true;

            }
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException("MysqlSource failed to connect to mysql database. " + e.getMessage());
        }

    }

    private Connection establishConn() throws ClassNotFoundException, IllegalAccessException, InstantiationException, SQLException {
        Class.forName("com.mysql.cj.jdbc.Driver").newInstance();
        String url = "jdbc:mysql://" + this.host + ":" + this.port + "/"
                + this.database + "?autoReconnect=true&useSSL=true";
        Connection connection = DriverManager.getConnection(url, this.username, this.password);
        // set to readonly to improve efficiency
        connection.setReadOnly(true);
        return connection;
    }

    private PreparedStatement getNextQuery() throws SQLException {
        boolean hasNextQuery;
        switch (batchByAttribute.getType()) {
            case INTEGER:
            case LONG:
            case TIMESTAMP:
                hasNextQuery = currentMin.longValue() <= max.longValue();
                break;
            case DOUBLE:
                hasNextQuery = currentMin.doubleValue() <= max.doubleValue();
                break;
            case STRING:
            case ANY:
            case BOOLEAN:
            default:
                throw new IllegalStateException("Unexpected value: " + schema.getAttribute(batchByAttribute.getName()).getType());
        }
        if (batchByAttribute.getType() == AttributeType.LONG || batchByAttribute.getType() == AttributeType.TIMESTAMP) {
            hasNextQuery = currentMin.longValue() <= max.longValue();
        }
        if (hasNextQuery) {
            PreparedStatement preparedStatement = this.connection.prepareStatement(generateSqlQuery());
            int curIndex = 1;
            if (column != null && keywords != null) {
                preparedStatement.setString(curIndex, keywords);
                curIndex += 1;
            }
            if (currentLimit != null) {
                preparedStatement.setLong(curIndex, currentLimit);
            }
            return preparedStatement;
        } else return null;
    }

    private void loadBatchColumnStats() {
        try {
            if (batchByAttribute != null && !batchByAttribute.getName().equals("")) {

                max = this.getBatchByBoundary("MAX");
                currentMin = this.getBatchByBoundary("MIN");

            }
        } catch (SQLException e) {
            e.printStackTrace();
            throw new RuntimeException("MysqlSource fail to load statistics on column `" + batchByAttribute.getName() + "`. " + e.getMessage());
        }

    }


    private Number getBatchByBoundary(String side) throws SQLException {
        Number result;
        PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT " + side + "(" + batchByAttribute.getName() + ") FROM " + table + ";");
        ResultSet resultSet = preparedStatement.executeQuery();
        resultSet.next();
        switch (schema.getAttribute(batchByAttribute.getName()).getType()) {
            case INTEGER:
                result = resultSet.getInt(1);
                break;
            case LONG:
                result = resultSet.getLong(1);
                break;
            case TIMESTAMP:
                result = resultSet.getTimestamp(1).getTime();
                break;
            case DOUBLE:
                result = resultSet.getDouble(1);
                break;
            case BOOLEAN:
            case STRING:
            case ANY:
            default:
                throw new IllegalStateException("Unexpected value: " + schema.getAttribute(batchByAttribute.getName()).getType());

        }

        // MAX is set to be 1 larger than the largest data, so that x < MAX can be outputted

        resultSet.close();
        preparedStatement.close();
        return result;
    }

    private void loadTableNames() throws SQLException {

        PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT table_name FROM information_schema.tables WHERE table_schema = ?;");

        preparedStatement.setString(1, database);
        ResultSet resultSet = preparedStatement.executeQuery();
        while (resultSet.next()) {
            tableNames.add(resultSet.getString(1));
        }

        resultSet.close();
        preparedStatement.close();

    }

    /**
     * close resultSet, preparedStatement and connection
     */
    @Override
    public void close() {
        try {
            if (resultSet != null) {
                resultSet.close();
            }
            if (currentPreparedStatement != null) {
                currentPreparedStatement.close();
            }
            if (connection != null) {
                connection.close();
            }
        } catch (SQLException e) {
            throw new RuntimeException("Mysql source fail to close. " + e.getMessage());
        }
    }

    /**
     * generate sql query string using the info provided by user. One of following
     * select * from TableName where 1 = 1 AND MATCH (ColumnName) AGAINST ( ? IN BOOLEAN MODE) LIMIT ?;
     * select * from TableName where 1 = 1 AND MATCH (ColumnName) AGAINST ( ? IN BOOLEAN MODE);
     * select * from TableName where 1 = 1 LIMIT ?;
     * select * from TableName where 1 = 1;
     *
     * @return string of sql query
     */
    private String generateSqlQuery() {
        // in sql prepared statement, table name cannot be inserted using preparedstatement.setString
        // so it has to be inserted here during sql query generation
        // this.table has to be verified to be existing in the given schema.
        String query = "\n" + "SELECT * FROM " + table + " where 1 = 1";
        // in sql prepared statement, column name cannot be inserted using preparedstatement.setString either
        if (column != null && keywords != null) {
            query += " AND MATCH(" + column + ") AGAINST (? IN BOOLEAN MODE)";
        }

        Number nextMin;
        boolean lastBatch;
        switch (batchByAttribute.getType()) {
            case INTEGER:
            case LONG:
            case TIMESTAMP:
                nextMin = currentMin.longValue() + interval;
                lastBatch = nextMin.longValue() >= max.longValue();
                break;
            case DOUBLE:
                nextMin = currentMin.doubleValue() + interval;
                lastBatch = nextMin.doubleValue() >= max.doubleValue();
                break;
            case BOOLEAN:
            case STRING:
            case ANY:
            default:
                throw new IllegalStateException("Unexpected value: " + schema.getAttribute(batchByAttribute.getName()).getType());
        }

        if (progressive) {
            query += " AND "
                    + batchByAttribute.getName() + " >= '"
                    + batchAttributeToString(currentMin) + "'"
                    + " AND "
                    + batchByAttribute.getName() +
                    (lastBatch ? (" <= '" + batchAttributeToString(max)) : (" < '" + batchAttributeToString(nextMin))) + "'";
        }
        currentMin = nextMin;

        if (currentLimit != null) {
            if (currentLimit < 0) {
                return null;
            }
            query += " LIMIT ?";
        }
        query += ";";
        System.out.println(query);
        return query;
    }


    private String batchAttributeToString(Number value) throws IllegalStateException {
        switch (batchByAttribute.getType()) {
            case LONG:
            case INTEGER:
            case DOUBLE:
                return String.valueOf(value);
            case TIMESTAMP:
                return new Timestamp(value.longValue()).toString();

            case BOOLEAN:
            case STRING:
            case ANY:
                throw new IllegalStateException("Unexpected value: " + schema.getAttribute(batchByAttribute.getName()).getType());
        }
        return null;
    }
}

File: common/workflow-operator/src/main/scala/org/apache/texera/amber/operator/source/sql/mysql/MySQLSourceOpExec.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.texera.amber.operator.source.sql.mysql

import io.vertx.core.Vertx
import io.vertx.sqlclient.{Pool, Row, SqlConnection, Transaction, Tuple => VxTuple}
import org.apache.texera.amber.core.executor.SourceOperatorExecutor
import org.apache.texera.amber.core.tuple.AttributeTypeUtils.{parseField, parseTimestamp}
import org.apache.texera.amber.core.tuple._
import org.apache.texera.amber.util.JSONUtils.objectMapper

import java.sql.Timestamp
import java.time.{LocalDate, LocalDateTime, LocalTime}
import java.util.concurrent.LinkedBlockingQueue
import scala.collection.mutable.ArrayBuffer

/**
  * MySQL source executor backed by the Vert.x MySQL client (Apache 2.0).
  *
  * The Vert.x client is reactive (`Future[RowSet]`), but the operator framework
  * pulls tuples via a synchronous `Iterator[TupleLike]`. We bridge the two with
  * a bounded [[LinkedBlockingQueue]]: the Vert.x event loop pushes rows as
  * they arrive and the iterator thread blocks on `take()`.
  *
  * This class intentionally does not extend [[org.apache.texera.amber.operator.source.sql.SQLSourceOpExec]]
  * because that base class is hard-wired to JDBC (`Connection`, `PreparedStatement`,
  * `ResultSet`). The query-building / keyword-search / progressive-batching logic
  * is re-implemented here against the Vert.x API.
  */
class MySQLSourceOpExec private[mysql] (descString: String) extends SourceOperatorExecutor {

  private val desc: MySQLSourceOpDesc =
    objectMapper.readValue(descString, classOf[MySQLSourceOpDesc])
  private var schema: Schema = _
  private var vertx: Vertx = _
  private var pool: Pool = _

  private val tableNames: ArrayBuffer[String] = ArrayBuffer()
  private var batchByAttribute: Option[Attribute] = None
  private var curLimit: Option[Long] = None
  private var curOffset: Option[Long] = None
  private var curLowerBound: Number = _
  private var upperBound: Number = _
  private var querySent: Boolean = false

  override def open(): Unit = {
    schema = desc.sourceSchema()
    curLimit = desc.limit
    curOffset = desc.offset

    vertx = Vertx.vertx()
    pool = MySQLConnUtil.createPool(
      vertx,
      desc.host,
      desc.port,
      desc.database,
      desc.username,
      desc.password
    )

    batchByAttribute =
      if (desc.progressive.getOrElse(false)) Option(schema.getAttribute(desc.batchByColumn.get))
      else None

    loadTableNames()
    if (!tableNames.contains(desc.table))
      throw new RuntimeException("Can't find the given table `" + desc.table + "`.")

    if (desc.progressive.getOrElse(false)) initBatchColumnBoundaries()
  }

  override def close(): Unit = {
    try if (pool != null) pool.close().toCompletionStage.toCompletableFuture.get()
    finally if (vertx != null) vertx.close().toCompletionStage.toCompletableFuture.get()
  }

  override def produceTuple(): Iterator[TupleLike] =
    new Iterator[TupleLike] {
      private var currentStream: Option[StreamingResult] = None
      private var cached: Option[Tuple] = None

      override def hasNext: Boolean = {
        if (cached.isDefined) return true
        cached = Option(advance())
        cached.isDefined
      }

      override def next(): TupleLike = {
        if (cached.isEmpty) cached = Option(advance())
        val t = cached.orNull
        cached = None
        if (t == null) throw new NoSuchElementException
        t
      }

      @scala.annotation.tailrec
      private def advance(): Tuple = {
        if (currentStream.isEmpty) {
          nextQuery() match {
            case Some((sql, params)) => currentStream = Some(startStream(sql, params))
            case None                => return null
          }
        }
        val stream = currentStream.get
        stream.nextRow() match {
          case Some(row) =>
            if (curOffset.exists(_ > 0)) {
              curOffset = curOffset.map(_ - 1)
              advance()
            } else {
              val tuple = buildTupleFromRow(row)
              curLimit.foreach(limit => if (limit > 0) curLimit = Some(limit - 1))
              tuple
            }
          case None =>
            currentStream = None
            advance()
        }
      }
    }

  // -- async → sync bridge ----------------------------------------------------

  private sealed trait StreamEvent
  private final case class DataRow(row: Row) extends StreamEvent
  private case object EndMarker extends StreamEvent
  private final case class ErrorMarker(err: Throwable) extends StreamEvent

  private final class StreamingResult(queue: LinkedBlockingQueue[StreamEvent]) {
    private var closed = false
    def nextRow(): Option[Row] = {
      if (closed) return None
      queue.take() match {
        case DataRow(r) => Some(r)
        case EndMarker =>
          closed = true
          None
        case ErrorMarker(err) =>
          closed = true
          throw new RuntimeException(err)
      }
    }
  }

  /**
    * Open a Vert.x cursor-backed stream for `sql` and push rows onto a blocking
    * queue. A transaction is required because Vert.x cursors only exist within
    * one.
    */
  private def startStream(sql: String, params: VxTuple): StreamingResult = {
    val queue = new LinkedBlockingQueue[StreamEvent]()

    pool
      .getConnection()
      .onSuccess { (conn: SqlConnection) =>
        conn
          .begin()
          .onSuccess { (tx: Transaction) =>
            conn
              .prepare(sql)
              .onSuccess { ps =>
                val stream = ps.createStream(256, params)
                stream.handler(row => queue.put(DataRow(row)))
                stream.endHandler { _ =>
                  tx.commit().onComplete(_ => {
                    conn.close()
                    queue.put(EndMarker)
                  })
                }
                stream.exceptionHandler { err =>
                  tx.rollback().onComplete(_ => {
                    conn.close()
                    queue.put(ErrorMarker(err))
                  })
                }
              }
              .onFailure { err =>
                tx.rollback().onComplete(_ => {
                  conn.close()
                  queue.put(ErrorMarker(err))
                })
              }
          }
          .onFailure { err =>
            conn.close()
            queue.put(ErrorMarker(err))
          }
      }
      .onFailure(err => queue.put(ErrorMarker(err)))

    new StreamingResult(queue)
  }

  // -- query building --------------------------------------------------------

  private def nextQuery(): Option[(String, VxTuple)] = {
    if (!hasNextQuery) return None
    val sql = new StringBuilder
    val params = VxTuple.tuple()

    sql ++= "SELECT * FROM " + desc.table + " WHERE 1 = 1"

    val keywordCol = desc.keywordSearchByColumn.orNull
    val keywords = desc.keywords.orNull
    if (desc.keywordSearch.getOrElse(false) && keywordCol != null && keywords != null) {
      val columnType = schema.getAttribute(keywordCol).getType
      if (columnType == AttributeType.STRING) {
        sql ++= " AND MATCH(" + keywordCol + ") AGAINST (? IN BOOLEAN MODE)"
        params.addString(keywords)
      } else throw new RuntimeException("Can't do keyword search on type " + columnType.toString)
    }

    if (desc.progressive.getOrElse(false) && desc.batchByColumn.isDefined && desc.interval > 0L)
      addBatchSlidingWindow(sql)

    if (curLimit.isDefined) {
      if (curLimit.get > 0) {
        sql ++= " LIMIT ?"
        params.addLong(curLimit.get)
      } else return None
    }

    if (!desc.progressive.getOrElse(false) && curOffset.isDefined) {
      sql ++= " OFFSET ?"
      params.addLong(curOffset.get)
    }

    sql ++= ";"
    Some((sql.result(), params))
  }

  private def hasNextQuery: Boolean =
    batchByAttribute match {
      case Some(attribute) =>
        attribute.getType match {
          case AttributeType.INTEGER | AttributeType.LONG | AttributeType.TIMESTAMP =>
            curLowerBound.longValue <= upperBound.longValue
          case AttributeType.DOUBLE =>
            curLowerBound.doubleValue <= upperBound.doubleValue
          case _ =>
            throw new IllegalArgumentException("Unexpected type: " + attribute.getType)
        }
      case None =>
        val has = !querySent
        querySent = true
        has
    }

  private def addBatchSlidingWindow(sql: StringBuilder): Unit = {
    var nextLowerBound: Number = null
    var isLast = false
    batchByAttribute match {
      case Some(attribute) =>
        attribute.getType match {
          case AttributeType.INTEGER | AttributeType.LONG | AttributeType.TIMESTAMP =>
            nextLowerBound = curLowerBound.longValue + desc.interval
            isLast = nextLowerBound.longValue >= upperBound.longValue
          case AttributeType.DOUBLE =>
            nextLowerBound = curLowerBound.doubleValue + desc.interval
            isLast = nextLowerBound.doubleValue >= upperBound.doubleValue
          case _ =>
            throw new IllegalArgumentException("Unexpected type: " + attribute.getType)
        }
        sql ++= " AND " + attribute.getName + " >= " + batchAttrToString(curLowerBound) +
          " AND " + attribute.getName +
          (if (isLast) " <= " + batchAttrToString(upperBound)
           else " < " + batchAttrToString(nextLowerBound))
      case None =>
        throw new IllegalArgumentException(
          "no valid batchByColumn to iterate: " + desc.batchByColumn.getOrElse("")
        )
    }
    curLowerBound = nextLowerBound
  }

  private def batchAttrToString(value: Number): String =
    batchByAttribute match {
      case Some(attribute) =>
        attribute.getType match {
          case AttributeType.LONG | AttributeType.INTEGER | AttributeType.DOUBLE =>
            String.valueOf(value)
          case AttributeType.TIMESTAMP =>
            "'" + new Timestamp(value.longValue).toString + "'"
          case _ =>
            throw new IllegalArgumentException("Unexpected type: " + attribute.getType)
        }
      case None =>
        throw new IllegalArgumentException(
          "No valid batchByColumn to iterate: " + desc.batchByColumn.getOrElse("")
        )
    }

  // -- bookkeeping queries ---------------------------------------------------

  private def loadTableNames(): Unit = {
    val rowSet = pool
      .preparedQuery("SELECT table_name FROM information_schema.tables WHERE table_schema = ?;")
      .execute(VxTuple.of(desc.database))
      .toCompletionStage
      .toCompletableFuture
      .get()
    val it = rowSet.iterator()
    while (it.hasNext) tableNames += it.next().getString(0)
  }

  private def initBatchColumnBoundaries(): Unit = {
    if (batchByAttribute.isDefined && desc.min.isDefined && desc.max.isDefined) {
      if (desc.min.get.equalsIgnoreCase("auto")) curLowerBound = fetchBoundary("MIN")
      else
        batchByAttribute.get.getType match {
          case AttributeType.TIMESTAMP => curLowerBound = parseTimestamp(desc.min.get).getTime
          case AttributeType.LONG      => curLowerBound = desc.min.get.toLong
          case _ =>
            throw new IllegalArgumentException(
              s"Unsupported type ${batchByAttribute.get.getType}"
            )
        }

      if (desc.max.get.equalsIgnoreCase("auto")) upperBound = fetchBoundary("MAX")
      else
        batchByAttribute.get.getType match {
          case AttributeType.TIMESTAMP => upperBound = parseTimestamp(desc.max.get).getTime
          case AttributeType.LONG      => upperBound = desc.max.get.toLong
          case _ =>
            throw new IllegalArgumentException(
              s"Unsupported type ${batchByAttribute.get.getType}"
            )
        }
    } else {
      throw new IllegalArgumentException(
        s"Missing required progressive configuration, $batchByAttribute, $desc.min or $desc.max."
      )
    }
  }

  private def fetchBoundary(side: String): Number =
    batchByAttribute match {
      case Some(attribute) =>
        val rowSet = pool
          .query("SELECT " + side + "(" + attribute.getName + ") FROM " + desc.table + ";")
          .execute()
          .toCompletionStage
          .toCompletableFuture
          .get()
        val it = rowSet.iterator()
        if (!it.hasNext) {
          val zero: Number = 0
          zero
        } else {
          val row = it.next()
          attribute.getType match {
            case AttributeType.INTEGER =>
              val n: Number = row.getInteger(0); n
            case AttributeType.LONG =>
              val n: Number = row.getLong(0); n
            case AttributeType.TIMESTAMP =>
              val ldt = row.getLocalDateTime(0)
              val n: Number = java.lang.Long.valueOf(Timestamp.valueOf(ldt).getTime); n
            case AttributeType.DOUBLE =>
              val n: Number = row.getDouble(0); n
            case _ =>
              throw new IllegalStateException("Unexpected value: " + attribute.getType)
          }
        }
      case None =>
        val zero: Number = 0
        zero
    }

  // -- row → Tuple -----------------------------------------------------------

  private def buildTupleFromRow(row: Row): Tuple = {
    val builder = Tuple.builder(schema)
    schema.getAttributes.iterator.zipWithIndex.foreach {
      case (attr, i) =>
        val raw = row.getValue(i)
        if (raw == null) builder.add(attr, null)
        else builder.add(attr, parseField(normalizeJavaTime(raw), attr.getType))
    }
    builder.build()
  }

  /**
    * Vert.x returns `LocalDateTime` / `LocalDate` / `LocalTime` for MySQL
    * temporal columns. Normalize to types that [[parseField]] understands.
    */
  private def normalizeJavaTime(value: Any): Any =
    value match {
      case ldt: LocalDateTime => Timestamp.valueOf(ldt)
      case ld: LocalDate      => ld.toString
      case lt: LocalTime      => lt.toString
      case other              => other
    }
}
